### ISSUE TITLE
cannot create OpenGL context,Check that the drivers and hardware support at least OpenGL4.3 (or re-run with -opengl3)

### ISSUE THREAD

**Issue by slnh312** [adamrehn/pixel-streaming-linux#2]
I have tried running unrealengine under ubuntu 18.04,it has an error an follow:
Warning:OpenGL is deprecated,please use Vilkan.
when I press ok button,a new error occured:
cannot create OpenGL context,Check that the drivers and hardware support at least OpenGL4.3 (or re-run with -opengl3)
I tried to follow the tips,but error occured.

**Comment by adamrehn** [comment]
Have you built the Unreal Engine from source using the [4.23.1-pixelstreaming](https://github.com/adamrehn/UnrealEngine/tree/4.23.1-pixelstreaming) branch of my fork of the Engine's repository? If so then it sounds like you need to ensure the correct drivers are installed for your graphics card.

**Comment by slnh312** [comment]
yes,I have built the UE from source using the 4.23.1-pixelstreaming branch of  your fork, the gpu drivers was installed successfully,and I can run the command : nvidia-smi,and with a correct reply:

nvidia-smi
Tue Mar  3 20:58:25 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 430.26       Driver Version: 430.26       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro RTX 5000     Off  | 00000000:00:0D.0 Off |                  Off |
| 33%   27C    P8     8W / 230W |      0MiB / 16125MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

**Comment by adamrehn** [comment]
Strange. Are you able to successfully run other OpenGL applications such as the `glxgears` demo from the [mesa-utils](https://packages.ubuntu.com/bionic/mesa-utils) package?

**Comment by slnh312** [comment]
I have just tried the glxgears demo from the mesa-utils package,it works correctly.

**Comment by slnh312** [comment]
My problem is same as follow：
https://www.epicgames.com/unrealtournament/forums/unreal-tournament-discussion/ut-development-bug-reports-qa/linux-troubleshooting/396371-using-gnome-and-cant-run-unreal

**Comment by adamrehn** [comment]
Hmmm, could you please post the output of the following commands:

```bash
# List our connected devices (we're interested in the entry for the NVIDIA GPU)
sudo lspci

# Print the contents of the X11 config file
sudo cat /etc/X11/xorg.conf
```

**Comment by slnh312** [comment]
sudo lspci
00:00.0 Host bridge: Intel Corporation 440FX - 82441FX PMC [Natoma] (rev 02)
00:01.0 ISA bridge: Intel Corporation 82371SB PIIX3 ISA [Natoma/Triton II]
00:01.1 IDE interface: Intel Corporation 82371SB PIIX3 IDE [Natoma/Triton II]
00:01.2 USB controller: Intel Corporation 82371SB PIIX3 USB [Natoma/Triton II] (rev 01)
00:01.3 Bridge: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 03)
00:02.0 VGA compatible controller: Cirrus Logic GD 5446
00:03.0 Ethernet controller: Red Hat, Inc. Virtio network device
00:04.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:05.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:06.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:07.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:08.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:09.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:0a.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:0b.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:0c.0 Communication controller: Red Hat, Inc. Virtio console
00:0d.0 VGA compatible controller: NVIDIA Corporation Device 1eb0 (rev a1)
00:0e.0 Audio device: NVIDIA Corporation Device 10f8 (rev a1)
00:0f.0 USB controller: NVIDIA Corporation Device 1ad8 (rev a1)
00:10.0 Serial bus controller [0c80]: NVIDIA Corporation Device 1ad9 (rev a1)
00:11.0 Unclassified device [00ff]: Red Hat, Inc. Virtio memory balloon
02:01.0 SCSI storage controller: Red Hat, Inc. Virtio block device

sudo cat /etc/X11/xorg.conf
# nvidia-xconfig: X configuration file generated by nvidia-xconfig
# nvidia-xconfig:  version 430.26

Section "ServerLayout"
    Identifier     "Layout0"
    Screen      0  "Screen0"
    InputDevice    "Keyboard0" "CoreKeyboard"
    InputDevice    "Mouse0" "CorePointer"
EndSection

Section "Files"
EndSection

Section "InputDevice"
    # generated from default
    Identifier     "Mouse0"
    Driver         "mouse"
    Option         "Protocol" "auto"
    Option         "Device" "/dev/psaux"
    Option         "Emulate3Buttons" "no"
    Option         "ZAxisMapping" "4 5"
EndSection

Section "InputDevice"
    # generated from default
    Identifier     "Keyboard0"
    Driver         "kbd"
EndSection

Section "Monitor"
    Identifier     "Monitor0"
    VendorName     "Unknown"
    ModelName      "Unknown"
    Option         "DPMS"
EndSection

Section "Device"
    Identifier     "Device0"
    Driver         "nvidia"
    VendorName     "NVIDIA Corporation"
    BusID          "PCI:0:13:0"
EndSection

Section "Screen"
    Identifier     "Screen0"
    Device         "Device0"
    Monitor        "Monitor0"
    DefaultDepth    24
    SubSection     "Display"
        Depth       24
    EndSubSection
EndSection

**Comment by adamrehn** [comment]
Well, the `BusID` value is correct (0:13:0 in decimal matches lspci's 00:0d:0 in hexadecimal), so it doesn't look like an incorrect X11 configuration is to blame.

**Comment by adamrehn** [comment]
Actually, there is one other thing worth checking before ruling out X11 as the cause. Could you please post the output of the following command:

```bash
# Force the Unreal Engine to render in offscreen mode instead of X11 mode
DISPLAY='' /path/to/UnrealEngine/Engine/Binaries/Linux/UE4Editor -opengl4 -nosplash
```

(Replace `/path/to/UnrealEngine` with the location of your Unreal Engine installation.)

**Comment by slnh312** [comment]
dear adamrehn,I am sorry to tell you so late that my system is running on Cloud with a GPU VM.hope it may help you .
I met some more questions in past few days.

I run the "glxinfo | grep OpenGL",it looks like running not in NVDIA GPU,

![1](https://user-images.githubusercontent.com/52794487/75966063-b3d7e580-5f04-11ea-83b0-8b64151e5817.png)

then I run the command "lspci | grep VGA",there are two gpus:
![2](https://user-images.githubusercontent.com/52794487/75966297-0fa26e80-5f05-11ea-9bed-ded291571df7.png)

1.I modify the /etc/X11/xorg.conf :  add the BusID ,then I can not start up the machine :
![20200305-171058(eSpace)](https://user-images.githubusercontent.com/52794487/75965819-517ee500-5f04-11ea-97fa-da1caeb9e089.png)

at last:
I hava tried your suggestion,it looks like running normally,thks to you,But I hope you can write an article that have more details  on Pixel Streaming in Linux containers ,thank you very much,

**Comment by adamrehn** [comment]
I already knew the system was a cloud VM the moment I saw references to `Red Hat, Inc. Virtio` in the `lspci` output, and had actually suggested checking the X11 config because I suspected as much from the start, having seen similar issues with VMs on AWS, which were solved using the instructions from [this repository](https://github.com/agisoft-llc/cloud-scripts).

The fact that the command to run offscreen rendering worked for you indicates that the problem was indeed related to the X11 configuration after all. Fortunately, this is not actually an issue for running Pixel Streaming inside Linux containers, since the container will use offscreen rendering mode by default unless you explicitly bind-mount the X11 socket from the host system and propagate the host's `DISPLAY` environment variable value. The issues you encountered were because the Unreal Engine was running directly on the host VM rather than inside a container.

My recommendation would be to develop your Pixel Streaming application using a local Linux machine, and use containers to deploy the packaged application to the cloud, which will circumvent these tricky X11 configuration issues and allow everything to run smoothly. I'm going to close this issue now, since the problem was caused by a configuration issue rather than a bug in the Linux Pixel Streaming implementation itself.
